sdks/csharp/sdk/AgonesSDK.cs

```diff
@@ -286,24 +324,74 @@ public async Task<Status> HealthAsync()
 		{
 			try
 			{
-				await healthStream.WriteAsync(new Empty());
+				await _healthStreamSemaphore.WaitAsync(ctoken);
+			}
+			catch (OperationCanceledException)
+			{
+				return Status.DefaultCancelled;
+			}
+
+			try
+			{
+				if (healthStream == null)
+				{
+					// Create a new stream if it's the first time we're being called or if the previous stream threw
+					// an exception.
+					healthStream = client.Health();
+				}
+
+				var writer = healthStream.RequestStream;
+				await writer.WriteAsync(new Empty());
 				return new Status(StatusCode.OK, "Health ping successful.");
 			}
 			catch (RpcException ex)
 			{
 				LogError(ex, "Unable to invoke the GameServer health check.");
+
+				if (healthStream != null)
+				{
+					try
+					{
+						// Best effort to clean up.
+						healthStream.Dispose();
+					}
+					catch
+					{
+						// Intentionally empty
+					}

```

Comment:
Similar note as above re empty catches. Debug log an option?

Reply:
The reasoning was that we're already logging when the exception is thrown above, and that consumers wouldn't really care if we failed to dispose the underlying broken stream, but I can also add a log here.